### PR TITLE
Authorization

### PR DESCRIPTION
**Pull request by MargaretTechworld**
In this PR, we have added CanCanCan to handle authorization in our project. The following changes have been made:

1. Installed CanCanCan gem.
2. Added a role column to the users table using migration.
3. Implemented post deleting functionality. Users can delete a post if it belongs to them or if they have an admin role.
4. Added a "Delete" button to the post view, which is only visible to authorized users.
5. Implemented comment deleting functionality. Users can delete a comment if it belongs to them or if they have an admin role.
6. Added a "Delete" button to the comment view, which is only visible to authorized users.

These changes ensure that only authorized users can delete posts and comments, providing a more secure and controlled environment. CanCanCan simplifies the authorization logic by defining abilities based on user roles.

Thank you!

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -21,6 +21,15 @@ def create
...
 
+  def destroy
+    @post = Post.find(params[:id])
+    @author = @post.author
+    @post.likes.destroy_all
+    @post.comments.destroy_all
+    @post.destroy
+    redirect_to user_posts_path(@author)
+  end
+
```
Comment (iambenkis) [inline app/controllers/posts_controller.rb:33]: - [ ] Good job adding authorization rules for the user , however you should include the link between the ability and the postController to ensure that the CanCanCan is used for Post authorization in Post Controller 👍 

```ruby
  class PostsController < ApplicationController
    load_and_authorize_resource

    ...
  end
```

File: app/models/ability.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    user ||= User.new # Guest user
+
+    can :read, :all
+
+    if user.role == 'admin'
+      can :manage, :all
+    else
+      can :destroy, Post, author_id: user.id
+      can :destroy, Comment, user_id: user.id
+    end
```
Comment (iambenkis) [inline app/models/ability.rb]: - [ ] Kudos adding the ability 👍 especially the admin management , however when a user is connected , he should be granted the access to manage all 👍 ,Kindly improve  that 💯 

```ruby
 # app/models/ability.rb

 class Ability
 include CanCan::Ability

     def initialize(user)

         ....

         if user.admin? 
         can :manage, :all
         else
         can :manage, Post, author: user  # owner of post can perform all view, update or destroy own post if Post.author == user
         can :manage, Comment, author: user 
         end
     end
 end
```

**Review by iambenkis — CHANGES_REQUESTED** [review]
#### Hi @MargaretTechworld , @BranBayou 

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!
You did a great job, keep up the good work 👍🏼

giphy


## Highlights

✔️  Add admin ability 

## Suggested changes
_Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

**Review by elisha2kyakpo1 — APPROVED** [review]
Hi 👋 team, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

✔️ You did well fixing the authorization code as requested by the previous reviewer.

Required 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**